File: src/calculus_derivative_rules_medium.py
"""
calculus - derivative_rules (medium)
Generated: 2026-02-22T05:02:45.892937
"""

from problem_utils import *

def generate():
    problem_type = choice(['power_rule', 'product_rule', 'quotient_rule', 'chain_rule', 'chain_product'])
    
    if problem_type == 'power_rule':
        # Standard polynomial derivative (1300-1400)
        num_terms = randint(2, 3)
        terms = []
        for _ in range(num_terms):
            coeff = nonzero(-8, 8)
            power = randint(2, 5)
            terms.append(coeff * x**power)
        
        # Add constant or linear term
        if choice([True, False]):
            terms.append(nonzero(-10, 10) * x)
        if choice([True, False]):
            terms.append(nonzero(-10, 10))
        
        expr = sum(terms)
        ans = diff(expr, x)
        
        return problem(
            question=f"\\text{{Find }} \\frac{{d}}{{dx}}\\left[{latex(expr)}\\right]",
            answer=ans,
            difficulty=(1300, 1400),
            topic="calculus/derivative_rules",
            solution=steps(
                f"Apply the power rule to each term of ${latex(expr)}$",
                f"For $x^n$, the derivative is $nx^{{n-1}}$",
                f"Result: ${latex(ans)}$"
            )
        )
    
    elif problem_type == 'product_rule':
        # Product rule with two simple functions (1400-1500)
        power1 = randint(2, 4)
        power2 = randint(2, 4)
        coeff1 = nonzero(-5, 5)
        coeff2 = nonzero(-5, 5)
        
        f = coeff1 * x**power1
        g = coeff2 * x**power2
        expr = f * g
        
        f_prime = diff(f, x)
        g_prime = diff(g, x)
        ans = simplify(diff(expr, x))
        
        return problem(
            question=f"\\text{{Find }} \\frac{{d}}{{dx}}\\left[{latex(expr)}\\right]",
            answer=ans,
            difficulty=(1400, 1500),
            topic="calculus/derivative_rules",
            solution=steps(
                f"Use the product rule: $(fg)' = f'g + fg'$",
                f"Let $f = {latex(f)}$ and $g = {latex(g)}$",
                f"Then $f' = {latex(f_prime)}$ and $g' = {latex(g_prime)}$",
                f"$(fg)' = ({latex(f_prime)})({latex(g)}) + ({latex(f)})({latex(g_prime)})$",
                f"Simplify: ${latex(ans)}$"
            )
        )
    
    elif problem_type == 'quotient_rule':
        # Quotient rule (1500-1600)
        power_num = randint(2, 3)
        power_den = randint(1, 2)
        coeff_num = nonzero(-6, 6)
        coeff_den = nonzero(-6, 6)
        
        numerator = coeff_num * x**power_num
        denominator = coeff_den * x**power_den
        
        num_prime = diff(numerator, x)
        den_prime = diff(denominator, x)
        
        ans = simplify(diff(numerator / denominator, x))
        
        return problem(
            question=f"\\text{{Find }} \\frac{{d}}{{dx}}\\left[\\frac{{{latex(numerator)}}}{{{latex(denominator)}}}\\right]",
            answer=ans,
            difficulty=(1500, 1600),
            topic="calculus/derivative_rules",
            solution=steps(
                f"Use the quotient rule: $\\left(\\frac{{f}}{{g}}\\right)' = \\frac{{f'g - fg'}}{{g^2}}$",
                f"Let $f = {latex(numerator)}$ and $g = {latex(denominator)}$",
                f"Then $f' = {latex(num_prime)}$ and $g' = {latex(den_prime)}$",
                f"$\\frac{{({latex(num_prime)})({latex(denominator)}) - ({latex(numerator)})({latex(den_prime)})}}{{({latex(denominator)})^2}}$",
                f"Simplify: ${latex(ans)}$"
            )
        )
    
    elif problem_type == 'chain_rule':
        # Chain rule with polynomial inside (1400-1500)
        outer_power = randint(2, 4)
        inner_coeff = nonzero(-5, 5)
        inner_const = nonzero(-8, 8)
        
        inner = inner_coeff * x + inner_const
        expr = inner**outer_power
        
        ans = simplify(diff(expr, x))
        
        return problem(
            question=f"\\text{{Find }} \\frac{{d}}{{dx}}\\left[{latex(expr)}\\right]",
            answer=ans,
            difficulty=(1400, 1500),
            topic="calculus/derivative_rules",
            solution=steps(
                f"Use the chain rule: $\\frac{{d}}{{dx}}[f(g(x))] = f'(g(x)) \\cdot g'(x)$",
                f"Let $u = {latex(inner)}$, so we have $u^{outer_power}$",
                f"Outer derivative: ${outer_power}u^{{{outer_power-1}}}$",
                f"Inner derivative: $\\frac{{du}}{{dx}} = {latex(diff(inner, x))}$",
                f"Multiply: ${outer_power}({latex(inner)})^{{{outer_power-1}}} \\cdot {latex(diff(inner, x))}$",
                f"Simplify: ${latex(ans)}$"
            )
        )
    
    else:  # chain_product
        # Combination of chain and product rule (1500-1600)
        power = randint(2, 3)
        coeff1 = nonzero(-4, 4)
        coeff2 = nonzero(-4, 4)
        const = nonzero(-5, 5)
        
        f = coeff1 * x**2
        g_inner = coeff2 * x + const
        g = g_inner**power
        expr = f * g
        
        ans = simplify(diff(expr, x))
        
        return problem(
            question=f"\\text{{Find }} \\frac{{d}}{{dx}}\\left[{latex(expr)}\\right]",
            answer=ans,
            difficulty=(1500, 1600),
            topic="calculus/derivative_rules",
            solution=steps(
                f"Use the product rule: $(fg)' = f'g + fg'$",
                f"Let $f = {latex(f)}$ and $g = {latex(g)}$",
                f"$f' = {latex(diff(f, x))}$",
                f"For $g'$, use the chain rule on ${latex(g)}$:",
                f"$g' = {power}({latex(g_inner)})^{{{power-1}}} \\cdot {latex(diff(g_inner, x))} = {latex(simplify(diff(g, x)))}$",
                f"Apply product rule and simplify: ${latex(ans)}$"
            )
        )

emit(generate())

File: src/calculus_derivative_rules_medium.jl
# calculus - derivative_rules (medium)
# Generated: 2026-03-08T20:39:08.699284

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("calculus/derivative_rules")
    
    problem_type = rand(1:8)
    
    if problem_type == 1
        # Power rule with polynomial (1200-1400)
        q = rand_poly(x, rand(3:5); coeff=(-8, 8))
        df = diff(q.expr, x)
        problem(
            question="Find the derivative: \\(\\frac{d}{dx}\\left[$(tex(q.expr))\\right]\\)",
            answer=df,
            difficulty=(1200, 1400),
            solution=steps(
                sol("Given", q.expr),
                "Apply the power rule to each term: \\(\\frac{d}{dx}[x^n] = nx^{n-1}\\)",
                sol("Answer", df)
            ),
            time=90
        )
    
    elseif problem_type == 2
        # Product rule (1400-1600)
        f = rand_linear(x; a=(-6, 6), b=(-10, 10))
        g = rand_quadratic(x; a=(-4, 4), b=(-8, 8), c=(-12, 12))
        product = expand(f.expr * g.expr)
        df = diff(f.expr, x)
        dg = diff(g.expr, x)
        result = expand(df * g.expr + f.expr * dg)
        problem(
            question="Find the derivative using the product rule: \\(\\frac{d}{dx}\\left[($(tex(f.expr)))($(tex(g.expr)))\\right]\\)",
            answer=result,
            difficulty=(1400, 1600),
            solution=steps(
                "Given: \\(f(x) = $(tex(f.expr))\\) and \\(g(x) = $(tex(g.expr))\\)",
                "Product rule: \\((fg)' = f'g + fg'\\)",
                sol("f'(x)", df),
                sol("g'(x)", dg),
                "Substitute: \\(($(tex(df)))($(tex(g.expr))) + ($(tex(f.expr)))($(tex(dg)))\\)",
                sol("Answer", result)
            ),
            time=180
        )
    
    elseif problem_type == 3
        # Quotient rule (1500-1700)
        numerator = rand_quadratic(x; a=(-5, 5), b=(-10, 10), c=(-15, 15))
        denom = rand_linear(x; a=(-6, 6), b=(-12, 12))
        dn = diff(numerator.expr, x)
        dd = diff(denom.expr, x)
        result = simplify((dn * denom.expr - numerator.expr * dd) / (denom.expr^2))
        problem(
            question="Find the derivative using the quotient rule: \\(\\frac{d}{dx}\\left[\\frac{$(tex(numerator.expr))}{$(tex(denom.expr))}\\right]\\)",
            answer=result,
            difficulty=(1500, 1700),
            solution=steps(
                "Given: \\(f(x) = $(tex(numerator.expr))\\) and \\(g(x) = $(tex(denom.expr))\\)",
                "Quotient rule: \\(\\left(\\frac{f}{g}\\right)' = \\frac{f'g - fg'}{g^2}\\)",
                sol("f'(x)", dn),
                sol("g'(x)", dd),
                "Substitute: \\(\\frac{($(tex(dn)))($(tex(denom.expr))) - ($(tex(numerator.expr)))($(tex(dd)))}{($(tex(denom.expr)))^2}\\)",
                sol("Simplified", result)
            ),
            time=240
        )
    
    elseif problem_type == 4
        # Chain rule with polynomial inside (1300-1500)
        inner = rand_linear(x; a=(-8, 8), b=(-15, 15))
        n = rand(3:6)
        outer_expr = inner.expr^n
        result = simplify(n * inner.expr^(n-1) * diff(inner.expr, x))
        problem(
            question="Find the derivative using the chain rule: \\(\\frac{d}{dx}\\left[($(tex(inner.expr)))^{$(n)}\\right]\\)",
            answer=result,
            difficulty=(1300, 1500),
            solution=steps(
                sol("Given", outer_expr),
                "Chain rule: \\(\\frac{d}{dx}[u^n] = nu^{n-1}\\cdot\\frac{du}{dx}\\) where \\(u = $(tex(inner.expr))\\)",
                sol("\\frac{du}{dx}", diff(inner.expr, x)),
                "Apply: \\($(n)($(tex(inner.expr)))^{$(n-1)} \\cdot $(tex(diff(inner.expr, x)))\\)",
                sol("Answer", result)
            ),
            time=150
        )
    
    elseif problem_type == 5
        # Product and chain rule combined (1600-1800)
        f = rand_linear(x; a=(-5, 5), b=(-10, 10))
        inner = rand_quadratic(x; a=(-4, 4), b=(-7, 7), c=(-10, 10))
        n = rand(2:4)
        g_expr = inner.expr^n
        df = diff(f.expr, x)
        dg = simplify(n * inner.expr^(n-1) * diff(inner.expr, x))
        result = simplify(df * g_expr + f.expr * dg)
        problem(
            question="Find the derivative: \\(\\frac{d}{dx}\\left[($(tex(f.expr)))($(tex(inner.expr)))^{$(n)}\\right]\\)",
            answer=result,
            difficulty=(1600, 1800),
            solution=steps(
                "Let \\(f(x) = $(tex(f.expr))\\) and \\(g(x) = ($(tex(inner.expr)))^{$(n)}\\)",
                "Product rule: \\((fg)' = f'g + fg'\\)",
                sol("f'(x)", df),
                "For \\(g'(x)\\), use chain rule:",
                sol("g'(x)", dg),
                "Combine: \\(($(tex(df)))($(tex(g_expr))) + ($(tex(f.expr)))($(tex(dg)))\\)",
                sol("Answer", result)
            ),
            time=240
        )
    
    elseif problem_type == 6
        # Derivative at a point (1200-1400)
        q = rand_quadratic(x; a=(-6, 6), b=(-10, 10), c=(-15, 15))
        point = randint(-8, 8)
        df = diff(q.expr, x)
        answer = substitute(df, x => point)
        problem(
            question="Find the derivative of \\(f(x) = $(tex(q.expr))\\) at \\(x = $(point)\\)",
            answer=answer,
            difficulty=(1200, 1400),
            solution=steps(
                sol("f(x)", q.expr),
                "Apply power rule:",
                sol("f'(x)", df),
                "Evaluate at \\(x = $(point)\\):",
                sol("f'($(point))", answer)
            ),
            time=120
        )
    
    elseif problem_type == 7
        # Higher order derivative (1400-1600)
        q = rand_poly(x, 4; coeff=(-6, 6))
        df1 = diff(q.expr, x)
        df2 = diff(df1, x)
        problem(
            question="Find the second derivative: \\(\\frac{d^2}{dx^2}\\left[$(tex(q.expr))\\right]\\)",
            answer=df2,
            difficulty=(1400, 1600),
            solution=steps(
                sol("f(x)", q.expr),
                "First derivative:",
                sol("f'(x)", df1),
                "Second derivative:",
                sol("f''(x)", df2)
            ),
            time=150
        )
    
    else
        # Chain rule with quotient (1700-1800)
        num = rand_linear(x; a=(-6, 6), b=(-12, 12))
        denom = rand_linear(x; a=(-5, 5), b=(-10, 10))
        n = rand(2:4)
        quotient = num.expr / denom.expr
        outer = quotient^n
        dn = diff(num.expr, x)
        dd = diff(denom.expr, x)
        inner_deriv = simplify((dn * denom.expr - num.expr * dd) / (denom.expr^2))
        result = simplify(n * quotient^(n-1) * inner_deriv)
        problem(
            question="Find the derivative: \\(\\frac{d}{dx}\\left[\\left(\\frac{$(tex(num.expr))}{$(tex(denom.expr))}\\right)^{$(n)}\\right]\\)",
            answer=result,
            difficulty=(1700, 1800),
            solution=steps(
                "Let \\(u = \\frac{$(tex(num.expr))}{$(tex(denom.expr))}\\)",
                "Chain rule: \\(\\frac{d}{dx}[u^{$(n)}] = $(n)u^{$(n-1)}\\cdot\\frac{du}{dx}\\)",
                "Use quotient rule for \\(\\frac{du}{dx}\\):",
                sol("\\frac{du}{dx}", inner_deriv),
                "Substitute back:",
                sol("Answer", result)
            ),
            time=300
        )
    end
end